### PR TITLE
Fix: Adjusted BPN validations on different operators

### PR DESCRIPTION
**Pull request by danielkenji01**
## WHAT
- Fixed **IS_ALL_OF** operator validation. It had the same behavior of **EQ** operator
- Fixed **IS_NONE_OF** operator validation. It had the same behavior of the **NEQ** operator
- Fixed **NEQ** operator validation. It was considering the opposite result of the IN operator, and it should be the opposite of the EQ operator
- Added a new rule when the BPN groups or assigned groups are empty, and if the operator is **NEQ** or **IS_NONE_OF**, it returns true instead of false

## WHY
In order to keep the correct behavior on BPNL validations

## FURTHER NOTES
Some minor adjusts reported by Sonar, including:
- Adjusted String format and casting

Closes #https://github.com/eclipse-tractusx/tractusx-edc/issues/1411
Closes #https://github.com/eclipse-tractusx/tractusx-edc/issues/775

### CODE REVIEW

File: edc-extensions/bpn-validation/bpn-validation-core/src/main/java/org/eclipse/tractusx/edc/validation/businesspartner/functions/BusinessPartnerGroupFunction.java [hunk 1]
```diff
@@ -120,44 +120,52 @@ public boolean evaluate(Operator operator, Object rightValue, Permission rule, P
...
-        if (rightOperand == null) {
+        if (rightOperand.isEmpty()) {
             return false;
         }
 
         //call evaluator function
         return OPERATOR_EVALUATOR_MAP.get(operator).apply(new BpnGroupHolder(assignedGroups, rightOperand));
     }
 
+    private boolean evaluateEmptyGroupsResponse(final Operator operator, final PolicyContext policyContext,
```
Comment (paullatzelsperger) [inline edc-extensions/bpn-validation/bpn-validation-core/src/main/java/org/eclipse/tractusx/edc/validation/businesspartner/functions/BusinessPartnerGroupFunction.java:148]: this method is quite obscure, from reading the signature it is not clear which combinations of parameters produce which result. Its fine to have a method to encapsulate error handling, but this is not whats happening here.
The method name is misleading as well. Please rewrite this, and a word of documentation can't hurt either.

File: edc-extensions/bpn-validation/bpn-validation-core/src/main/java/org/eclipse/tractusx/edc/validation/businesspartner/functions/BusinessPartnerGroupFunction.java [hunk 2]
```diff
@@ -120,44 +120,52 @@ public boolean evaluate(Operator operator, Object rightValue, Permission rule, P
...
             return Arrays.asList(tokens);
         }
         if (rightValue instanceof Collection<?>) {
             return ((Collection<?>) rightValue).stream().map(Object::toString).toList();
         }
 
-        context.reportProblem(format("Right operand expected to be either String or a Collection, but was " + rightValue.getClass()));
-        return null;
+        context.reportProblem(format("Right operand expected to be either String or a Collection, but was %s", rightValue.getClass()));
+        return Collections.emptyList();
```
Comment (paullatzelsperger) [inline edc-extensions/bpn-validation/bpn-validation-core/src/main/java/org/eclipse/tractusx/edc/validation/businesspartner/functions/BusinessPartnerGroupFunction.java:168]: is there a reason for this? how can you then distinguish between an empty array `"[]"` and an invalid right operand?

File: edc-extensions/bpn-validation/bpn-validation-core/src/test/java/org/eclipse/tractusx/edc/validation/businesspartner/functions/BusinessPartnerGroupFunctionTest.java [hunk 3]
```diff
@@ -82,14 +85,14 @@ void setUp() {
     void evaluate_noParticipantAgentOnContext() {
         reset(context);
         assertThat(function.evaluate(EQ, "test-group", createPermission(EQ, List.of()), context)).isFalse();
-        verify(context).reportProblem(eq("ParticipantAgent not found on PolicyContext"));
+        verify(context).reportProblem("ParticipantAgent not found on PolicyContext");
```
Comment (paullatzelsperger) [inline edc-extensions/bpn-validation/bpn-validation-core/src/test/java/org/eclipse/tractusx/edc/validation/businesspartner/functions/BusinessPartnerGroupFunctionTest.java:88]: please don't just change unrelated code. Supplying a matcher is good practice IMO, because it increases expressivity.

File: edc-extensions/bpn-validation/bpn-validation-core/src/test/java/org/eclipse/tractusx/edc/validation/businesspartner/functions/BusinessPartnerGroupFunctionTest.java [hunk 4]
```diff
@@ -82,14 +85,14 @@ void setUp() {
...
     }
 
     @ParameterizedTest(name = "Invalid operator {0}")
     @ArgumentsSource(InvalidOperatorProvider.class)
     @DisplayName("Invalid operators, expect report in policy context")
     void evaluate_invalidOperator(Operator invalidOperator) {
-        when(context.getContextData(eq(ParticipantAgent.class))).thenReturn(new ParticipantAgent(Map.of(), Map.of()));
+        when(context.getContextData(ParticipantAgent.class)).thenReturn(new ParticipantAgent(Map.of(), Map.of()));
```
Comment (paullatzelsperger) [inline edc-extensions/bpn-validation/bpn-validation-core/src/test/java/org/eclipse/tractusx/edc/validation/businesspartner/functions/BusinessPartnerGroupFunctionTest.java:95]: see above.

File: edc-extensions/bpn-validation/bpn-validation-core/src/test/java/org/eclipse/tractusx/edc/validation/businesspartner/functions/BusinessPartnerGroupFunctionTest.java [hunk 5]
```diff
@@ -98,7 +101,7 @@ void evaluate_invalidOperator(Operator invalidOperator) {
     @DisplayName("Right-hand operand is not String or Collection<?>")
     void evaluate_rightOperandNotStringOrCollection() {
         when(store.resolveForBpn(TEST_BPN)).thenReturn(StoreResult.success(List.of("test-group")));
-        when(context.getContextData(eq(ParticipantAgent.class))).thenReturn(new ParticipantAgent(Map.of(), Map.of(PARTICIPANT_IDENTITY, TEST_BPN)));
+        when(context.getContextData(ParticipantAgent.class)).thenReturn(new ParticipantAgent(Map.of(), Map.of(PARTICIPANT_IDENTITY, TEST_BPN)));
```
Comment (paullatzelsperger) [inline edc-extensions/bpn-validation/bpn-validation-core/src/test/java/org/eclipse/tractusx/edc/validation/businesspartner/functions/BusinessPartnerGroupFunctionTest.java:104]: ditto

File: edc-extensions/bpn-validation/bpn-validation-core/src/test/java/org/eclipse/tractusx/edc/validation/businesspartner/functions/BusinessPartnerGroupFunctionTest.java [hunk 6]
```diff
@@ -117,29 +120,51 @@ void evaluate_rightOperandNotStringOrCollection() {
     void evaluate_validOperator(String ignored, Operator operator, List<String> assignedBpn, boolean expectedOutcome) {
 
         var allowedGroups = List.of(TEST_GROUP_1, TEST_GROUP_2);
-        when(context.getContextData(eq(ParticipantAgent.class))).thenReturn(new ParticipantAgent(Map.of(), Map.of(PARTICIPANT_IDENTITY, TEST_BPN)));
+        when(context.getContextData(ParticipantAgent.class)).thenReturn(new ParticipantAgent(Map.of(), Map.of(PARTICIPANT_IDENTITY, TEST_BPN)));
```
Comment (paullatzelsperger) [inline edc-extensions/bpn-validation/bpn-validation-core/src/test/java/org/eclipse/tractusx/edc/validation/businesspartner/functions/BusinessPartnerGroupFunctionTest.java:123]: ditto

File: edc-extensions/bpn-validation/bpn-validation-core/src/test/java/org/eclipse/tractusx/edc/validation/businesspartner/functions/BusinessPartnerGroupFunctionTest.java [hunk 7]
```diff
@@ -117,29 +120,51 @@ void evaluate_rightOperandNotStringOrCollection() {
...
         when(store.resolveForBpn(TEST_BPN)).thenReturn(StoreResult.success(assignedBpn));
         assertThat(function.evaluate(operator, allowedGroups, createPermission(operator, allowedGroups), context)).isEqualTo(expectedOutcome);
     }
 
-    @Test
-    void evaluate_noEntryForBpn() {
-        var operator = NEQ;
+    @EnumSource(value = Operator.class, names = {"NEQ", "IS_NONE_OF"})
+    @ParameterizedTest
+    void evaluate_noEntryForBpnWithNegativeOperator_shouldBeTrue(final Operator operator) {
```
Comment (paullatzelsperger) [inline edc-extensions/bpn-validation/bpn-validation-core/src/test/java/org/eclipse/tractusx/edc/validation/businesspartner/functions/BusinessPartnerGroupFunctionTest.java:130]: please no `final`s in method signatures, local vars, etc. `final` is only allowed in field declarations. check out our coding guidelines.

File: edc-extensions/bpn-validation/bpn-validation-core/src/test/java/org/eclipse/tractusx/edc/validation/businesspartner/functions/BusinessPartnerGroupFunctionTest.java [hunk 8]
```diff
@@ -117,29 +120,51 @@ void evaluate_rightOperandNotStringOrCollection() {
...
         var allowedGroups = List.of(TEST_GROUP_1, TEST_GROUP_2);
-        when(context.getContextData(eq(ParticipantAgent.class))).thenReturn(new ParticipantAgent(Map.of(), Map.of(PARTICIPANT_IDENTITY, TEST_BPN)));
+        when(context.getContextData(ParticipantAgent.class)).thenReturn(new ParticipantAgent(Map.of(), Map.of(PARTICIPANT_IDENTITY, TEST_BPN)));
         when(store.resolveForBpn(TEST_BPN)).thenReturn(StoreResult.notFound("foobar"));
 
-        assertThat(function.evaluate(operator, allowedGroups, createPermission(operator, allowedGroups), context)).isFalse();
+        assertTrue(function.evaluate(operator, allowedGroups, createPermission(operator, allowedGroups), context));
```
Comment (paullatzelsperger) [inline edc-extensions/bpn-validation/bpn-validation-core/src/test/java/org/eclipse/tractusx/edc/validation/businesspartner/functions/BusinessPartnerGroupFunctionTest.java:135]: please don't change existing code without good reason. we use assertJ assertions to keep consistency. goes for the whole file.

**Review by paullatzelsperger — CHANGES_REQUESTED** [review]
TL;DR: don't just change unrelated code, don't use `final` except for field declarations

File: edc-extensions/bpn-validation/bpn-validation-core/src/main/java/org/eclipse/tractusx/edc/validation/businesspartner/functions/BusinessPartnerGroupFunction.java (hunk 2 above)
Comment (danielkenji01) [reply edc-extensions/bpn-validation/bpn-validation-core/src/main/java/org/eclipse/tractusx/edc/validation/businesspartner/functions/BusinessPartnerGroupFunction.java]: It was more of a suggestion from Sonar to return an empty list instead of a null value, then I changed the validation after the method to verify if the list is empty.

File: edc-extensions/bpn-validation/bpn-validation-core/src/main/java/org/eclipse/tractusx/edc/validation/businesspartner/functions/BusinessPartnerGroupFunction.java (hunk 1 above)
Comment (danielkenji01) [reply edc-extensions/bpn-validation/bpn-validation-core/src/main/java/org/eclipse/tractusx/edc/validation/businesspartner/functions/BusinessPartnerGroupFunction.java]: I refactored the method and added a documentation to be more clear

**Comment by danielkenji01** [comment]
> TL;DR: don't just change unrelated code, don't use `final` except for field declarations

Thank you for the comments, I followed them and adjusted the code.

File: edc-extensions/bpn-validation/bpn-validation-core/src/main/java/org/eclipse/tractusx/edc/validation/businesspartner/functions/BusinessPartnerGroupFunction.java [hunk 9]
```diff
@@ -120,20 +121,13 @@ public boolean evaluate(Operator operator, Object rightValue, Permission rule, P
...
         var bpn = participantAgent.getIdentity();
         var groups = store.resolveForBpn(bpn);
 
-        // BPN not found in database
-        if (groups.failed()) {
-            policyContext.reportProblem(groups.getFailureDetail());
-            return false;
-        }
-        var assignedGroups = groups.getContent();
+        var errorMessage = evaluateEmptyGroups(groups, operator, bpn);
```
Comment (paullatzelsperger) [inline edc-extensions/bpn-validation/bpn-validation-core/src/main/java/org/eclipse/tractusx/edc/validation/businesspartner/functions/BusinessPartnerGroupFunction.java:127]: results should not be passed into methods. ever. they have a fluent API, so please use it. at this point though, the `evaluateEmptyGroups` method is not needed and its content can be inlined here.

File: edc-extensions/bpn-validation/bpn-validation-core/src/main/java/org/eclipse/tractusx/edc/validation/businesspartner/functions/BusinessPartnerGroupFunction.java [hunk 10]
```diff
@@ -143,20 +137,57 @@ public boolean evaluate(Operator operator, Object rightValue, Permission rule, P
...
+     */
+    private String evaluateEmptyGroups(StoreResult<List<String>> groups, Operator operator, String bpn) {
+        var isNegativeOperator = NEGATIVE_OPERATORS.contains(operator);
+
+        if (isNegativeOperator) {
+            return null;
+        }
+
+        // BPN not found in database
+        if (groups.failed()) {
```
Comment (paullatzelsperger) [inline edc-extensions/bpn-validation/bpn-validation-core/src/main/java/org/eclipse/tractusx/edc/validation/businesspartner/functions/BusinessPartnerGroupFunction.java:167]: like I remarked above, `Result`s should be passed into and evaluate in methods. The code flow of this method and its caller can be simplified/unified and made more readable.

File: edc-extensions/bpn-validation/bpn-validation-core/src/test/java/org/eclipse/tractusx/edc/validation/businesspartner/functions/BusinessPartnerGroupFunctionTest.java [hunk 11]
```diff
@@ -122,24 +124,46 @@ void evaluate_validOperator(String ignored, Operator operator, List<String> assi
...
+    @ParameterizedTest
+    void evaluate_noEntryForBpnWithNegativeOperator_shouldBeTrue(Operator operator) {
+        var allowedGroups = List.of(TEST_GROUP_1, TEST_GROUP_2);
+        when(context.getContextData(eq(ParticipantAgent.class))).thenReturn(new ParticipantAgent(Map.of(), Map.of(PARTICIPANT_IDENTITY, TEST_BPN)));
+        when(store.resolveForBpn(TEST_BPN)).thenReturn(StoreResult.notFound("foobar"));
+
+        assertThat(function.evaluate(operator, allowedGroups, createPermission(operator, allowedGroups), context)).isTrue();
+    }
+
+    @EnumSource(value = Operator.class, names = {"NEQ", "IS_NONE_OF"})
```
Comment (paullatzelsperger) [inline edc-extensions/bpn-validation/bpn-validation-core/src/test/java/org/eclipse/tractusx/edc/validation/businesspartner/functions/BusinessPartnerGroupFunctionTest.java:137]: Test coverage is insufficient. you're only testing one scenario, i.e. left-hand operand is empty, right-hand operand is non-empty.
What about `[a,b,c] NEQ []`, `[a,b,c] IS_NONE_OF []`? same goes for the "positive" operator test.

**Review by paullatzelsperger — CHANGES_REQUESTED** [review]

File: edc-extensions/bpn-validation/bpn-validation-core/src/main/java/org/eclipse/tractusx/edc/validation/businesspartner/functions/BusinessPartnerGroupFunction.java (hunk 9 above)
Comment (danielkenji01) [reply edc-extensions/bpn-validation/bpn-validation-core/src/main/java/org/eclipse/tractusx/edc/validation/businesspartner/functions/BusinessPartnerGroupFunction.java]: I simplified the validation and removed the method.

File: edc-extensions/bpn-validation/bpn-validation-core/src/test/java/org/eclipse/tractusx/edc/validation/businesspartner/functions/BusinessPartnerGroupFunctionTest.java (hunk 11 above)
Comment (danielkenji01) [reply edc-extensions/bpn-validation/bpn-validation-core/src/test/java/org/eclipse/tractusx/edc/validation/businesspartner/functions/BusinessPartnerGroupFunctionTest.java:136]: I added the test scenarios.

File: edc-extensions/bpn-validation/bpn-validation-core/src/main/java/org/eclipse/tractusx/edc/validation/businesspartner/functions/BusinessPartnerGroupFunction.java [hunk 12]
```diff
@@ -120,20 +120,23 @@ public boolean evaluate(Operator operator, Object rightValue, Permission rule, P
...
-            policyContext.reportProblem(groups.getFailureDetail());
-            return false;
-        }
-        var assignedGroups = groups.getContent();
+            errorMessage = groups.getFailureDetail();
 
         // BPN was found, but it does not have groups assigned.
-        if (assignedGroups.isEmpty()) {
-            policyContext.reportProblem("No groups were assigned to BPN " + bpn);
+        } else if (groups.getContent().isEmpty()) {
```
Comment (paullatzelsperger) [inline edc-extensions/bpn-validation/bpn-validation-core/src/main/java/org/eclipse/tractusx/edc/validation/businesspartner/functions/BusinessPartnerGroupFunction.java:134]: this code can be simplified, since an empty result is only an error if the operator is "positive", so your if-else cascades should reflect that in the form of a nested if-else: 
```java
if(result.isFailed()){
  policyContext.reportProblem(result.getFailureDetail());
  return false;
}
var groups = result.getGroups();
if (isPositiveOperator && groups.isEmpty()) {
    policyContext.reportProblem("No groupse assigned...");
    return false;
}

```

**Review by paullatzelsperger — CHANGES_REQUESTED** [review]

File: edc-extensions/bpn-validation/bpn-validation-core/src/main/java/org/eclipse/tractusx/edc/validation/businesspartner/functions/BusinessPartnerGroupFunction.java (hunk 12 above)
Comment (danielkenji01) [reply edc-extensions/bpn-validation/bpn-validation-core/src/main/java/org/eclipse/tractusx/edc/validation/businesspartner/functions/BusinessPartnerGroupFunction.java:134]: Regarding the first question: even though I assign an error message to the variable, I just report the error if it is a positive operator, as it is validated on the line 138. 
About the other test scenarios, I will add them here.

Comment (paullatzelsperger) [reply edc-extensions/bpn-validation/bpn-validation-core/src/main/java/org/eclipse/tractusx/edc/validation/businesspartner/functions/BusinessPartnerGroupFunction.java:134]: no, i'm saying that an empty assigned group is _not_ an error, because if the policy states ` eq []`, then a BPN that has no groups assigned to it would satisfy that condition. 

Generally though, even if the code were semantically correct, the control flow is too complicated, there are duplicated fragments.

Comment (danielkenji01) [reply edc-extensions/bpn-validation/bpn-validation-core/src/main/java/org/eclipse/tractusx/edc/validation/businesspartner/functions/BusinessPartnerGroupFunction.java:134]: So, in this case, would you agree that both _groups.failed()_ and _groups.getContent().isEmpty()_ validations wouldn't make sense anymore? Since in both cases, we could consider the result as an empty list, then validate directly on the operator evaluators. 

For example, we would have:
- `[] eq []` = true
- `[] isAllOf []` = true
- `[] isAnyOf []` = false
- `[] in []` = false

Comment (paullatzelsperger) [reply edc-extensions/bpn-validation/bpn-validation-core/src/main/java/org/eclipse/tractusx/edc/validation/businesspartner/functions/BusinessPartnerGroupFunction.java:134]: nope, that's incorrect. there is a fundamental difference between a `Result.failure` and an empty list. A `failure` could be caused e.g. by a database error or something. That is an error situation. An empty list is not an error situation, because a particular BPN might simply not have any groups assigned. There could still be constraints that evaluate to `true`. 

I also disagree with your examples, because `[] isAnyOf []` should be true. Any element on the left side (Which is none) is equal to any element of the right side (which is also none). And `[] in []` is also true, because the entire list on the left side (which is empty) is contained fully in the right side (also being empty).

If you have no apples, and I ask you to give me all your apples, and I end up with no apples - did you give me everything you had?

Comment (danielkenji01) [reply edc-extensions/bpn-validation/bpn-validation-core/src/main/java/org/eclipse/tractusx/edc/validation/businesspartner/functions/BusinessPartnerGroupFunction.java:134]: Okay, so just this validation `groups.getContent().isEmpty()` would not make sense anymore? 
In regard to the validations from each operator, I will check and fix any inconsistent behavior.

Comment (danielkenji01) [reply edc-extensions/bpn-validation/bpn-validation-core/src/main/java/org/eclipse/tractusx/edc/validation/businesspartner/functions/BusinessPartnerGroupFunction.java]: @paullatzelsperger I don't know why, but I'm not being able to re-request a review, as you asked me to do after every conversation solved on this PR.

File: edc-extensions/bpn-validation/bpn-validation-core/src/test/java/org/eclipse/tractusx/edc/validation/businesspartner/functions/BusinessPartnerGroupFunctionTest.java [hunk 13]
```diff
@@ -115,31 +117,45 @@ void evaluate_rightOperandNotStringOrCollection() {
...
-        assertThat(function.evaluate(operator, allowedGroups, createPermission(operator, allowedGroups), context)).isFalse();
+        assertThat(function.evaluate(operator, allowedGroups, createPermission(operator, allowedGroups), context)).isTrue();
+        verify(context, never()).reportProblem("foobar");
+    }
+
+    @ArgumentsSource(OperatorForEmptyGroupsProvider.class)
+    @ParameterizedTest
+    void evaluate_groupsAssignedButNoGroupsSentToEvaluate(Operator operator, List<String> assignedBpnGroups,
+                                                          boolean expectedOutcome) {
+        var allowedGroups = List.<String>of();
```
Comment (paullatzelsperger) [inline edc-extensions/bpn-validation/bpn-validation-core/src/test/java/org/eclipse/tractusx/edc/validation/businesspartner/functions/BusinessPartnerGroupFunctionTest.java:153]: nit:
```suggestion
        List<String> allowedGroups = List.of();
```

**Review by paullatzelsperger — CHANGES_REQUESTED** [review]

File: edc-extensions/bpn-validation/bpn-validation-core/src/main/java/org/eclipse/tractusx/edc/validation/businesspartner/functions/BusinessPartnerGroupFunction.java [hunk 14]
```diff
@@ -148,35 +142,48 @@ public boolean evaluate(Operator operator, Object rightValue, Permission rule, P
...
-        return !evaluateIn(bpnGroupHolder);
+        return !evaluateEquals(bpnGroupHolder);
     }
 
     private Boolean evaluateEquals(BpnGroupHolder bpnGroupHolder) {
         return bpnGroupHolder.allowedGroups.equals(bpnGroupHolder.assignedGroups);
     }
 
+    private boolean evaluateIsAnyOf(BpnGroupHolder bpnGroupHolder) {
+        if (bpnGroupHolder.allowedGroups.isEmpty() && bpnGroupHolder.assignedGroups.isEmpty()) {
```
Comment (paullatzelsperger) [inline edc-extensions/bpn-validation/bpn-validation-core/src/main/java/org/eclipse/tractusx/edc/validation/businesspartner/functions/BusinessPartnerGroupFunction.java:172]: not necessary, this is covered by the stream operation.

**Review by paullatzelsperger — CHANGES_REQUESTED** [review]

File: edc-extensions/bpn-validation/bpn-validation-core/src/main/java/org/eclipse/tractusx/edc/validation/businesspartner/functions/BusinessPartnerGroupFunction.java (hunk 14 above)
Comment (danielkenji01) [reply edc-extensions/bpn-validation/bpn-validation-core/src/main/java/org/eclipse/tractusx/edc/validation/businesspartner/functions/BusinessPartnerGroupFunction.java:172]: It doesn't cover, that why I had to add this separate empty lists validation. I can try to find another why to incorporate this validation with the stream() methods, but I still don't know how. I accept any suggestions that you might have for this case.

**Review by paullatzelsperger — APPROVED** [review]
